shared/Microsoft.Extensions.StackTrace.Sources/StackFrame/StackTraceHelper.cs

```diff
@@ -119,5 +144,105 @@ internal static MethodDisplayInfo GetMethodDisplayString(MethodBase method)
 
             return methodDisplayInfo;
         }
+
+        private static bool ShowInStackTrace(MethodBase method)
+        {
+            Debug.Assert(method != null);
+
+            // Don't show any methods marked with the StackTraceHiddenAttribute
+            // https://github.com/dotnet/coreclr/pull/14652
+            foreach (var attibute in method.CustomAttributes)
+            {
+                // internal Attribute, match on name
+                if (attibute.AttributeType.Name == "StackTraceHiddenAttribute")
+                {
+                    return false;
+                }
+            }
+
+            var type = method.DeclaringType;
+            if (type == null)
+            {
+                return true;
+            }
+
+            foreach (var attibute in type.CustomAttributes)
+            {
+                // internal Attribute, match on name
+                if (attibute.AttributeType.Name == "StackTraceHiddenAttribute")
+                {
+                    return false;
+                }
+            }
+
+            // Fallbacks for runtime pre-StackTraceHiddenAttribute
+            if (type == typeof(ExceptionDispatchInfo) && method.Name == "Throw")
+            {
+                return false;
+            }
+            else if (type == typeof(TaskAwaiter) || 
+                type == typeof(TaskAwaiter<>) ||
+                type == typeof(ConfiguredTaskAwaitable.ConfiguredTaskAwaiter) ||
+                type == typeof(ConfiguredTaskAwaitable<>.ConfiguredTaskAwaiter))
+            {
+                switch (method.Name)
+                {
+                    case "HandleNonSuccessAndDebuggerNotification":
+                    case "ThrowForNonSuccess":
+                    case "ValidateEnd":
+                    case "GetResult":
+                        return false;
+                }
+            }
+            else if (type.FullName == "System.ThrowHelper")
```

Comment:
This type is a real type that shows up in real source code, no? Shouldn't it be left in the call stack?

I totally agree removing the "junk" that doesn't show up in anyone's source code (as in, not even in .NET's source code), but I'm less certain about legitimate method calls.

Reply:
> This type is a real type that shows up in real source code, no? Shouldn't it be left in the call stack?

Its just a type to move the throw out of the generic code to reduce instantiation code bloat. So it essentially turns the call stack back into looking like an inline throw (as its just an implementation detail)

e.g. from
```
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
  at System.ThrowHelper.ThrowInvalidOperationException_InvalidOperation_EnumFailedVersion()
  at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
```
to
```
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
  at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
```

It will also be suppressed in future `Exception.ToString()` https://github.com/dotnet/coreclr/pull/14652 so its just matching ASP.NET's stack frame processing to match that.

Reply:
Also they will also get hidden when running on latest 2.1.x as it checking for the `StackTraceHiddenAttribute` so is just bringing Framework inline

Reply:
Yeah I know `ThrowHelper` is cruft, but it's explicit cruft that an engineer chose to put in their code. I personally strongly dislike using such helpers because it *does* affect the stack trace and can potentially inhibit debugging tools from doing their jobs. That is, where did this ArgNullException *really* come from?

But, I don't see why the ThrowHelper needs to be special cased here. Won't https://github.com/dotnet/coreclr/pull/14652/files#diff-49215540f67fdd24553e4844693948b4R45 resolve it on the new runtime? I'm not sure I like the idea of ASP.NET Core checking for an odd implementation detail of .NET (and especially one which I 100% disagree with 😄).

Reply:
Removed the removal of ThrowHelper

Reply:
> I personally strongly dislike using such helpers because it does affect the stack trace...

[Slides 19-22](https://www.ageofascent.com/2017/11/05/perfromance-dotnet-core-2-corestart-conference/) 😉

Reply:
Yeah inlining is a real problem in some cases. I just don't like working around "the system's" limitations. You want exceptions to be thrown from where the error happens, but you don't want to because then it won't inline, so you have to write helpers for it instead. So I say go fix .NET! 😄